### DISCUSSION TITLE
Sealed Render method in Avalonia.Controls.Image

### DISCUSSION THREAD

**Discussion by mvegaca** [AvaloniaUI/Avalonia#11496] [Q&A]
I am working on a Avalonia project. We have a class that inherits from Avalonia.Controls.Image and overrides the Render method. In one of the commits to Avalonia's master branch, the Avalonia team has marked the Render method of the Image control as sealed, making not possible to override it. Does anyone know the reason of that? Is there any alternative method to perform a custom drawing operation on an Image control? I have also tried using public new void Render(DrawingContext context), but my method is not being executed.

Visual Studio error when I try to override the Render method.

![image](https://github.com/AvaloniaUI/Avalonia/assets/5556784/ed3ac5bf-72ca-475e-a844-3f0718123f4e)

Render method on Master.

![image](https://github.com/AvaloniaUI/Avalonia/assets/5556784/0aa3f364-7e61-4aff-999d-c0b865ddbe5a)

Thank you all so much in advanced!

**Comment by timunie** [comment]
I guess it's better to either base on `Control` for your own Image or use a `RenderTargetBitmap` or `WriteableBitmap` as source. 

The reaseon for making Render sealed is, that this gives more flexibility in the 11.x timeline for planned changes which would not be possible otherwise I guess.

**Reply by mvegaca** [reply]
I've update the control to extend from Control instead of Image and it works I think in this case would be better to extend from Image because the resulting object is an image and should habe image properties like Stretch, StretchDirection, etc.

**Comment by maxkatz6** [comment]
What exactly are you trying to do?

- Draw border on top of image? Use Border control together with Image control.
- Draw custom drawing content (lines, shapes...) on top of the image? You either can combine it with Avalonia shape controls on top of an image or draw it to the WriteableBitmap/RenderTargetBitmap and render it as a second image on top. Or you can create a custom control and override Render there. And once again put it on top of the Image class.

In other words, keep Image to do its purpose - drawing itself. And any other drawings to make either with the original source before hands or in separated control on top of it.

**Reply by maxkatz6** [reply]
And about reasoning, see this issue https://github.com/AvaloniaUI/Avalonia/issues/10066

**Comment by Baltasarq** [comment]
I have a simple component that draws a line or a bar chart. In order to support that, I used Canvas and overrode *Canvas.Render()*.

```c#
public class Chart: Canvas {
        /// <summary>Select the type of chart.</summary>
        public enum ChartType { Lines, Bars }

        // ...

        public override void Render(DrawingContext ctx)
        {
            base.Render( ctx );
            this.Draw();
        }

        /// <summary>Redraws the chart</summary>
        public void Draw()
        {
            this.ClearCanvas();
            
            // Frame
            this.DrawRectangle(
                // ...
        }
}
```

I cannot do that anymore, since **Canvas**.*Render()* has been marked sealed. My problem is that I cannot find a suitable workaround to that. The best approach seems to be to add an event handler to *GotFocus*, which is unfortunately insufficient.

What can I do to obtain the same behaviour as before?

Thanks in advance,
-- Baltasar

**Reply by jp2masa** [reply]
`Canvas` is a panel, not actually a canvas for drawing (the name is a bit misleading, it comes from WPF). Instead, you should inherit from `Control`, for example.

**Comment by Baltasarq** [comment]
I'm afraid that doesn't work. **Control** apparently does not have a *Background* property, or what is more important, a *Children* property to add shapes (Rectangles, Lines, Strings, and so forth), inside it. I cannot have a SetLeft() or SetTop() functions to set the coordinates of each shape.

Or maybe I'm missing something big here.

**Reply by jp2masa** [reply]
I think you can just add a child control filling the canvas and override `Render` on that control.

**Comment by Baltasarq** [comment]
I'm afraid I don't get it. The current [source code](https://github.com/baltasarq/DemoAvalonia) for [Chart](https://github.com/Baltasarq/DemoAvalonia/blob/main/DemoAvalonia/UI/Chart.cs) follows the behaviour below.

```c#
public class Chart: Canvas {
        // ....
        private void DrawRectangle(Pen pen, int x, int y, int width, int height)
        {
            var rect = new Rectangle {
                Width = width, Height = height,
                Stroke = pen.Brush,
                StrokeThickness = pen.Thickness,
            };
            
            SetLeft( rect, x );
            SetTop( rect, y );
            this.Children.Add( rect );
        }
}
```

So, if I add a **Control** as a child of **Canvas**, I'd have the same problem: I wouldn't be able to add rectangles, lines... I mean, I assume you mean that, say, a **Line** would be inside of **Control**, and then the latter would be inside the **Canvas**.

I tried the contrary: adding a **Canvas** as a child of the **Chart**, now being a **Control**. Unfortunately, **Control** does not have the *Children* property, so I used *VisualChildren* to add a new **Canvas** object.

```c#
public class Chart: Control {
    public Chart()
    {
            // Create canvas
            this.canvas = new Canvas();
            this.VisualChildren.Add( this.canvas ); 
    }

    public override void Render(DrawingContext context)
    {
            base.Render( context );
            this.Draw();
    }

    // ...
}
```

Unfortunately, when rendering now I got the exception:

    Unhandled exception. System.InvalidOperationException: Visual was invalidated during the render pass

Any other idea?

Thanks,
-- Baltasar

**Reply by workgroupengineering** [reply]
See this for draw line
 https://github.com/workgroupengineering/AvaloniaVS/blob/master/AvaloniaVS.Shared%2FViews%2FGridLines.cs

```xaml
<Grid>
<c:GridLines/>
<Canvas/>
</Grid>
```

**Comment by Baltasarq** [comment]
Interesting. Why *OnRender()* instead of *Render()*, provided it inherits from **Control**? Is this code for an older version?

Anyway, if I can draw with **DrawingContext**, then I can get rid of **Canvas**.

If this **Canvas** is just intended for child *controls*, then it is definitely misnamed. Perhaps it should be a **CanvasPanel**. Or at least, the examples in the [documentation](https://docs.avaloniaui.net/docs/reference/controls/canvas) should not suggest to indeed draw lines and rectangles.

```c#
<Canvas Background="AliceBlue" Margin="20">
  <Rectangle Fill="Red" Height="100" Width="100" Margin="10"/>
  <Rectangle Fill="Blue" Height="100" Width="100" Opacity="0.5"
             Canvas.Left="50" Canvas.Top="20"/>
  <Rectangle Fill="Green" Height="100" Width="100" 
             Canvas.Left="60" Margin="40" Canvas.Top="40"/>
  <Rectangle Fill="Orange" Height="100" Width="100" 
             Canvas.Right="70" Canvas.Bottom="60"/>
</Canvas>
```

**Comment by Baltasarq** [comment]
I have a few answers. I want the graph to be redrawn from outside the control, for instance because the user changed something (lines for bars, a different font...). So now I need something to invalidate the control and force *Render()* to be called. The problem is, of course, the **DrawingContext**, which only appears as a parameter of the *Render()* method.

```c#
class Graph: Control {
        public override void Render(DrawingContext graphics)
        {
            base.Render( graphics );
            this.Draw( graphics );
        }

        public void Draw()
        {
            this.InvalidateVisual();
        }

        /// <summary>Redraws the chart</summary>
        private void Draw(DrawingContext graphics)
        {
            this.graphics = graphics;
            this.ClearCanvas();
        }

        // ...
}
```

Is this *InvalidateVisual()* appropriate to force a re-render and therefore have *Draw(graphics)* called?

**Reply by jp2masa** [reply]
> If this Canvas is just intended for child controls, then it is definitely misnamed. Perhaps it should be a CanvasPanel. Or at least, the examples in the [documentation](https://docs.avaloniaui.net/docs/reference/controls/canvas) should not suggest to indeed draw lines and rectangles.

`Rectangle` is a `Control`.

I was suggesting something like this:

```xaml
<Canvas>
    <c:MyCustomControl />
    <Rectangle Fill="Blue" Height="100" Width="100" Opacity="0.5" Canvas.Left="50" Canvas.Top="20"/>
    <Rectangle Fill="Green" Height="100" Width="100"  Canvas.Left="60" Margin="40" Canvas.Top="40"/>
</Canvas>
```

Where `MyCustomControl` overrides `Render`.

**Comment by Baltasarq** [comment]
> Rectangle is a Control.

Great, but I don't think that's very important; it enforces the finder of this docs further in that indeed this is the appropriate "control" for drawing shapes.

I think that an example in which a TextBox and a Button (for example) are set with fixed coordinates would illustrate more the true nature and use of Canvas.

```xml
<Canvas>
    <TextBox Canvas.Left="50" Canvas.Top="20" />
    <Button Canvas.Left="50" Canvas.Top="150" Content="Ok" />
</Canvas>
```

> I was suggesting something like this:

I've nearly everything sorted out... except drawing text in vertical. You see, I was adding a label and rotating it through *transform()* 270 degrees. I've been looking for the answer but was unable to find anything forcing **DrawingContext**.*DrawString()* to draw text upwards.

Any clue?
-- Baltasar

**Reply by timunie** [reply]
No not really. It's fir most Apps not a good idea to position TextBox or Button by coordinates. Thus we don't want to show that in the docs. Grid or DockPanel are most of the time better suited

**Reply by Baltasarq** [reply]
Sure, I wouldn't do that myself.

But then... what is **Canvas** for? I mean, is it a **Panel**, used to arrange controls by their absolute position, or is it a control used to draw shapes?

**Reply by maxkatz6** [reply]
It is a Panel that positions elements by their Canvas.Left/Top/Right/Bottom properties.

**Comment by Baltasarq** [comment]
Mmm... I think that **Canvas** was actually thought to draw shapes, at least for what I can find in [the docs about drawing shapes in Avalonia](https://docs.avaloniaui.net/docs/guides/graphics-and-animation/graphics-and-animations).

In that link you can find the following example:

```xml
<Canvas Background="Yellow" Width="300" Height="400">
    <Rectangle Fill="Blue" Width="63" Height="41" Canvas.Left="40" Canvas.Top="31">
        <Rectangle.OpacityMask>
            <LinearGradientBrush StartPoint="0%,0%" EndPoint="100%,100%">
                <LinearGradientBrush.GradientStops>
                    <GradientStop Offset="0" Color="Black"/>
                    <GradientStop Offset="1" Color="Transparent"/>
                </LinearGradientBrush.GradientStops>
            </LinearGradientBrush>
        </Rectangle.OpacityMask>     
    </Rectangle>
   <!-- ... -->
</Canvas>
```

**Reply by timunie** [reply]
Canvas is a Panel not needed for shapes. But you always need a kind of panel for your children and Canvas is well suited here. But you can use any other Panel as pointed out earlier

**Reply by Baltasarq** [reply]
Sure, it's just that I think the docs are misguiding.

**Reply by timunie** [reply]
Hmmm 🤔 not sure what to do here. There is another section about panels. If you know a better wording feel free to open a PR to improve the docs. But I don't really see how.

**Comment by Baltasarq** [comment]
This is important to me now: I've been looking for the answer but was unable to find anything about forcing **DrawingContext**.*DrawString()* to draw text upwards.

**Reply by timunie** [reply]
You may need to push a transform before. Don't remember the exact syntax

**Reply by Baltasarq** [reply]
I need help for that. It seems you need to give the exact coordinates (not the control, since, well, there is not one), and then transform through a rotation.

So I'd need to:

1. Find the coordinates and size of where the string is going to be drawn.
2. Call transform appropriately.
3. "Pop" it when it is over.

I'm not a WPF developer.

**Comment by Baltasarq** [comment]
Can anybody help me here?

I'm struggling with matrix and transform, but don't see how to advance here.

```c#
private void DrawString(
                    Font font, IBrush color, int x, int y, string msg,
                    bool vertical = false)
    {
        // ...        
        
        // Formatted text
        var text = new FormattedText(
                            msg,
                            CultureInfo.CurrentCulture,
                            flow,
                            typeface, 
                            font.Size,
                            color );
        
        // Vertical
        if ( vertical ) {
            var m = new Matrix();
            m.Transform( new Point( x, y ) );
            this.Graphics.PushTransform( m );
            //lbl.RenderTransform = new RotateTransform( 270.0 );  // Old code with canvas
        }

        this.Graphics.DrawText( text, new Point( x, y ) );
    }
```

I know I have to probably put something inside the parens when creating the **Matrix**. Unfortunately, the params don't ring a bell on me (*shiftx*, or *skewx*?). I also need a way to indicate that the transformation to do is to rotate the text 270º.

**Reply by timunie** [reply]
```cs
var matrix = new RotateTransform(270).Value; 
```

**Reply by Baltasarq** [reply]
Thanks!! It works!!
